pkg/reconciler/pipelinerun/pipelinerun_test.go

```diff
@@ -0,0 +1,268 @@
+					}},
+			},
+			shouldSign: false,
+		},
+		{
+			name: "taskruns completed",
+			pr: &v1beta1.PipelineRun{
+				ObjectMeta: metav1.ObjectMeta{
+					Annotations: map[string]string{},
+				},
+				Status: v1beta1.PipelineRunStatus{
+					Status: duckv1beta1.Status{
+						Conditions: []apis.Condition{{Type: apis.ConditionSucceeded}},
+					},
+					PipelineRunStatusFields: v1beta1.PipelineRunStatusFields{
+						TaskRuns: map[string]*v1beta1.PipelineRunTaskRunStatus{
+							"taskrun1": {
+								PipelineTaskName: "task1",
+								Status: &v1beta1.TaskRunStatus{
+									TaskRunStatusFields: v1beta1.TaskRunStatusFields{
+										CompletionTime: &metav1.Time{},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			taskruns: []*v1beta1.TaskRun{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "taskrun1",
+						Annotations: map[string]string{
+							"chains.tekton.dev/signed": "true",
+						},
+					},
+					Status: v1beta1.TaskRunStatus{
+						TaskRunStatusFields: v1beta1.TaskRunStatusFields{
+							CompletionTime: &v1.Time{Time: time.Date(1995, time.December, 24, 6, 12, 12, 24, time.UTC)},
+						},
+					},
+				},
+			},
+			shouldSign: true,
+			wantErr:    false,
+		},
+		{
+			name: "taskruns not yet completed",
+			pr: &v1beta1.PipelineRun{
+				ObjectMeta: metav1.ObjectMeta{
+					Annotations: map[string]string{},
+				},
+				Status: v1beta1.PipelineRunStatus{
+					Status: duckv1beta1.Status{
+						Conditions: []apis.Condition{{Type: apis.ConditionSucceeded}},
+					},
+					PipelineRunStatusFields: v1beta1.PipelineRunStatusFields{
+						TaskRuns: map[string]*v1beta1.PipelineRunTaskRunStatus{
+							"taskrun1": {
+								PipelineTaskName: "task1",
+								Status: &v1beta1.TaskRunStatus{
+									TaskRunStatusFields: v1beta1.TaskRunStatusFields{
+										CompletionTime: &metav1.Time{},
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+			taskruns: []*v1beta1.TaskRun{
+				{
+					ObjectMeta: metav1.ObjectMeta{
+						Name: "taskrun1",
+					},
+				},
+			},
+			shouldSign: false,
+			wantErr:    true,
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			signer := &mockSigner{}
+			ctx, _ := rtesting.SetupFakeContext(t)
+			c := fakepipelineclient.Get(ctx)
+
+			r := &Reconciler{
+				PipelineRunSigner: signer,
+				Pipelineclientset: c,
+			}
+
+			// Create mock taskruns
+			for _, tr := range tt.taskruns {
+				err := test.CreateTektonObject(ctx, r.Pipelineclientset, objects.NewTaskRunObject(tr))
+				if err != nil {
+					t.Errorf("Unable to create mock taskrun: %s", tr.Name)
+				}
+			}
+
+			if err := r.ReconcileKind(ctx, tt.pr); err != nil && !tt.wantErr {
+				t.Errorf("Reconciler.handlePipelineRun() error = %v", err)
+			}
+			if signer.Signed != tt.shouldSign {
+				t.Errorf("Reconciler.handlePipelineRun() signed = %v, wanted %v", signer.Signed, tt.shouldSign)
+			}
+		})
+	}
+}
+
+type mockSigner struct {
```

Comment:
Should we share this in an internal package with the taskrun reconciler?

Reply:
Agreed, fixed.